### PR TITLE
[IMP] sheetview: prevent useless SET_VIEWPORT_OFFSET

### PR DESCRIPTION
**Pull request by hokolomopo**
## [IMP] sheetview: prevent useless SET_VIEWPORT_OFFSET:

When the user scrolls at the boundaries of a sheet, a lot of
SET_VIEWPORT_OFFSET commands are dispatched and trigger new renders,
even if the viewport offset is not actually changed.

This commit adds an allowDispatch in the sheetview plugin to prevent the
dispatch of SET_VIEWPORT_OFFSET that won't scroll any viewport.

## [IMP] sheetview: remove useless Object.values()

The SheetView plugin was doing Object.values() on arrays, which was
both useless and allocated some memory for new arrays.

This commit removes the useless Object.values() calls.

Task: : [3846620](https://www.odoo.com/web#id=3846620&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/plugins/ui_stateful/sheetview.ts [hunk 1]
```diff
@@ -640,6 +643,18 @@ export class SheetViewPlugin extends UIPlugin {
     return CommandResult.Success;
   }
 
+  private checkViewportsWillChange({ offsetX, offsetY }: SetViewportOffsetCommand) {
```
Comment (anhe-odoo) [inline src/plugins/ui_stateful/sheetview.ts:646]: Nitpickest : I don't like this name but I don't have any better idea ... Maybe `CheckViewportsChange(s)` or `CheckIfViewportsWillChange` ?

File: src/plugins/ui_stateful/sheetview.ts [hunk 2]
```diff
@@ -687,7 +702,7 @@ export class SheetViewPlugin extends UIPlugin {
   private setSheetViewOffset(offsetX: Pixel, offsetY: Pixel) {
     const sheetId = this.getters.getActiveSheetId();
     const { maxOffsetX, maxOffsetY } = this.getMaximumSheetOffset();
-    Object.values(this.getSubViewports(sheetId)).forEach((viewport) =>
+    this.getSubViewports(sheetId).forEach((viewport) =>
```
Comment (anhe-odoo) [inline src/plugins/ui_stateful/sheetview.ts:705]: :+1:

File: src/types/commands.ts [hunk 3]
```diff
@@ -1097,6 +1097,7 @@ export const enum CommandResult {
   Readonly = "Readonly",
   InvalidViewportSize = "InvalidViewportSize",
   InvalidScrollingDirection = "InvalidScrollingDirection",
+  NoViewportScroll = "NoViewportScroll",
```
Comment (anhe-odoo) [inline src/types/commands.ts:1100]: `ViewportScrollLimitsReached` or `ViewportWontScrool` ? But, once again, it's a total nitpick and maybe your name is the best one :)

**Review by anhe-odoo — COMMENTED** [review]
LGTM :)

File: src/types/commands.ts (hunk 3 above)
Comment (hokolomopo) [reply src/types/commands.ts:1100]: ViewportScrollLimitsReached sounds way better, I didn't have any good idea when coding. Thanks ! ;)

**Comment by rrahir** [comment]
@robodoo r+

**Comment by rrahir** [comment]
@robodoo rebase-ff

**Pull request closed without merging**